pdns/dnsdistdist/doq.cc

```diff
@@ -0,0 +1,963 @@
+    if (written == QUICHE_ERR_DONE) {
+      return;
+    }
+
+    if (written < 0) {
+      return;
+    }
+    // FIXME pacing (as send_info.at should tell us when to send the packet) ?
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    sock.sendTo(reinterpret_cast<const char*>(out.data()), written, conn.d_peer);
+  }
+}
+
+std::unique_ptr<CrossProtocolQuery> getDOQCrossProtocolQueryFromDQ(DNSQuestion& dnsQuestion, bool isResponse)
+{
+  if (!dnsQuestion.ids.doqu) {
+    throw std::runtime_error("Trying to create a DoQ cross protocol query without a valid DoQ unit");
+  }
+
+  auto unit = std::move(dnsQuestion.ids.doqu);
+  if (&dnsQuestion.ids != &unit->ids) {
+    unit->ids = std::move(dnsQuestion.ids);
+  }
+
+  unit->ids.origID = dnsQuestion.getHeader()->id;
+
+  if (!isResponse) {
+    if (unit->query.data() != dnsQuestion.getMutableData().data()) {
+      unit->query = std::move(dnsQuestion.getMutableData());
+    }
+  }
+  else {
+    if (unit->response.data() != dnsQuestion.getMutableData().data()) {
+      unit->response = std::move(dnsQuestion.getMutableData());
+    }
+  }
+
+  return std::make_unique<DOQCrossProtocolQuery>(std::move(unit), isResponse);
+}
+
+/*
+   We are not in the main DoQ thread but in the DoQ 'client' thread.
+*/
+static void processDOQQuery(DOQUnitUniquePtr&& doqUnit)
+{
+  const auto handleImmediateResponse = [](DOQUnitUniquePtr&& unit, [[maybe_unused]] const char* reason) {
+    DEBUGLOG("handleImmediateResponse() reason=" << reason);
+    auto conn = getConnection(unit->dsc->df->d_server_config->d_connections, unit->serverConnID);
+    handleResponse(*unit->dsc->df, *conn, unit->streamID, unit->response);
+    unit->ids.doqu.reset();
+  };
+
+  auto& ids = doqUnit->ids;
+  ids.doqu = std::move(doqUnit);
+  auto& unit = ids.doqu;
+  uint16_t queryId = 0;
+  ComboAddress remote;
+
+  try {
+
+    remote = unit->ids.origRemote;
+    DOQServerConfig* dsc = unit->dsc;
+    auto& holders = dsc->holders;
+    ClientState& clientState = *dsc->clientState;
+
+    if (unit->query.size() < sizeof(dnsheader)) {
+      ++dnsdist::metrics::g_stats.nonCompliantQueries;
+      ++clientState.nonCompliantQueries;
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      auto* dnsHeader = reinterpret_cast<struct dnsheader*>(unit->query.data());
+      dnsHeader->rcode = RCode::ServFail;
+      dnsHeader->qr = true;
+      unit->response = std::move(unit->query);
+
+      handleImmediateResponse(std::move(unit), "DoQ non-compliant query");
+      return;
+    }
+
+    ++clientState.queries;
+    ++dnsdist::metrics::g_stats.queries;
+    unit->ids.queryRealTime.start();
+
+    {
+      /* don't keep that pointer around, it will be invalidated if the buffer is ever resized */
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+      auto* dnsHeader = reinterpret_cast<struct dnsheader*>(unit->query.data());
```

Comment:
We should watch out for unaligned access here

Reply:
I have looked into this, and also other places where it might happen, which led to a pretty big commit: https://github.com/PowerDNS/pdns/commit/ca0346f268979da7ebeb2cd1ae4aff2395b08b63 I think we might want to make this a separate PR.

Reply:
I pushed a new commit taking care of potential unaligned accesses in the DoQ code. I'll open a new PR for the existing code once we have merged this one.